### PR TITLE
fix: multi content types support

### PR DESCRIPTION
**Pull request by TommoLeedsy**
This was a limitation of the previous implementation where only 1 content type per route was recorded. This has now been rectified in the PR and content types are attached to body and return types. This should have been rectified #26 but this was not a know issue until it was integrated into a more complex system.

### CODE REVIEW

File: examples/basic/types/error.go
Comment (Will-Mann-16) [inline examples/basic/types/error.go:1]: Was this used? If not can it be deleted.

File: examples/with-multi-content-types/deletePost.go [hunk 1]
```diff
@@ -25,19 +25,18 @@ func DeletePostJSON(c *gin.Context) {
...
-	err := c.ShouldBindUri(postURI)
-	if err != nil {
-		c.JSON(http.StatusBadRequest, types.Error{
-			Error:   "cannot bind uri",
-			Details: err.Error(),
-		})
-		return
+	switch c.ContentType() {
+	case "application/json":
+		{
```
Comment (Will-Mann-16) [inline examples/with-multi-content-types/deletePost.go:30]: These work well in JS I get it, but personally I don't like them here. Go FMT does us no favours and frankly I'm not a fan.

File: examples/with-multi-content-types/main.go [hunk 2]
```diff
@@ -10,16 +10,11 @@ import (
...
-	r.POST("/posts/yaml", CreatePostYAML)
-	r.PUT("/posts/:id/json", UpdatePostJSON)
-	r.PUT("/posts/:id/yaml", UpdatePostYAML)
-	r.DELETE("/posts/:id/json", DeletePostJSON)
-	r.DELETE("/posts/:id/yaml", DeletePostYAML)
+	r.GET("/posts", GetPosts)
+	r.GET("/posts/:id", GetPost)
+	r.POST("/posts", CreatePost)
+	r.PUT("/posts/:id", UpdatePost)
+	r.DELETE("/posts/:id", DeletePost)
```
Comment (Will-Mann-16) [inline examples/with-multi-content-types/main.go:18]: Can there be a readme for this example explaining what it is for?

File: inputs/gin/parseFunction.go [hunk 3]
```diff
@@ -335,11 +330,10 @@ func parseFunction(s *astra.Service, funcTraverser *astTraversal.FunctionTravers
 					currRoute, err = funcBuilder.ExpressionResult().Build(func(route *astra.Route, params []any) (*astra.Route, error) {
 						field := astra.ParseResultToField(params[0].(astTraversal.Result))
 
-						route.BodyType = "form"
-
-						route.Body = append(route.Body, astra.Param{
-							IsBound: true,
-							Field:   field,
+						route.Body = append(route.Body, astra.BodyParam{
+							ContentType: "form",
```
Comment (Will-Mann-16) [inline inputs/gin/parseFunction.go:334]: Is the content type here form or `application/x-www-form-urlencoded` or `multipart/form-data`?

Comment (Will-Mann-16) [reply inputs/gin/parseFunction.go:334]: Realising I put this in...

File: types.go [hunk 4]
```diff
@@ -46,6 +45,17 @@ type Param struct {
...
+	Name        string `json:"name,omitempty" yaml:"name,omitempty"`
+	Field       Field  `json:"type,omitempty" yaml:"type,omitempty"`
+	ContentType string `json:"contentType,omitempty" yaml:"contentType,omitempty"`
+	IsRequired  bool   `json:"isRequired,omitempty" yaml:"isRequired,omitempty"`
+	IsArray     bool   `json:"isArray,omitempty" yaml:"isArray,omitempty"`
+	IsMap       bool   `json:"isMap,omitempty" yaml:"isMap,omitempty"`
+
+	IsBound bool `json:"isBound,omitempty" yaml:"isBound,omitempty"` // I.e. is a struct reference
+}
+
```
Comment (Will-Mann-16) [inline types.go:58]: Did the struct embedding idea not work?

File: outputs/openapi/generate.go [hunk 5]
```diff
@@ -124,23 +128,21 @@ func Generate(filePath string) astra.ServiceFunction {
...
 				mediaType.Schema = mapFieldToSchema(bindingType, returnType.Field)
 
-				var content map[string]MediaType
-				if mediaType.Schema.Type != "" || mediaType.Schema.Ref != "" {
-					content = map[string]MediaType{
-						endpoint.ContentType: mediaType,
+				statusCode := strconv.Itoa(returnType.StatusCode)
+				if _, set := operation.Responses[statusCode]; !set {
+					operation.Responses[statusCode] = Response{
+						Description: "",
```
Comment (Will-Mann-16) [inline outputs/openapi/generate.go:137]: Nit: if empty do we need to define them here? Mainly for `Links` and `Description`, completely see the argument for `Content` as an empty map is nil

**Review by Will-Mann-16 — CHANGES_REQUESTED** [review]

File: inputs/gin/parseFunction.go (hunk 3 above)
Comment (TommoLeedsy) [reply inputs/gin/parseFunction.go:334]: Yeah this is me just moving around your code, tbh I have actually missed this so need to fix that anyway

File: outputs/openapi/generate.go (hunk 5 above)
Comment (TommoLeedsy) [reply outputs/openapi/generate.go:137]: I see what you saying but even if I initialise the `Content` map here but don't add any items it's still considered empty. I problem I was having before was that I was putting empty structs into the map.

File: types.go (hunk 4 above)
Comment (TommoLeedsy) [reply types.go:58]: Unfortunately not, I would have mean initialising the embedded struct separately every time, I can do that but actually I think we should review what is in each param struct anyway.

File: examples/basic/types/error.go
Comment (TommoLeedsy) [reply examples/basic/types/error.go:1]: This was accidentally committed and has been removed.

File: examples/with-multi-content-types/deletePost.go (hunk 1 above)
Comment (TommoLeedsy) [reply examples/with-multi-content-types/deletePost.go:30]: Hmmm, I quite liked them, I know there quiet large but im not sure that's a problem here?

File: inputs/gin/parseFunction.go (hunk 3 above)
Comment (TommoLeedsy) [reply inputs/gin/parseFunction.go]: This should now be solved, there is a slight problem with the implementation but I don't think that's within the scope of this PR to fix, Its when multiple structs are embedded in a single struct.

File: examples/with-multi-content-types/main.go (hunk 2 above)
Comment (TommoLeedsy) [reply examples/with-multi-content-types/main.go:18]: I think there already is, what do you want me to add?

File: examples/with-gorm/openapi.generated.yaml [hunk 6]
```diff
@@ -147,35 +147,72 @@ paths:
...
-            additionalProperties: {}
-        gorm.DeletedAt:
-            type: object
-            properties:
-                Time:
-                    $ref: '#/components/schemas/time.Time'
-                Valid:
-                    type: boolean
-        gorm.Model:
+        "":
```
Comment (Will-Mann-16) [inline examples/with-gorm/openapi.generated.yaml:150]: This doesn't look right to me

File: examples/with-cli/openapi.generated.yaml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 openapi: 3.0.0
 info:
-    title: Example API with CLI
```
Comment (Will-Mann-16) [inline examples/with-cli/openapi.generated.yaml:3]: Why was this removed

File: examples/with-multi-content-types/main.go (hunk 2 above)
Comment (Will-Mann-16) [reply examples/with-multi-content-types/main.go:18]: Explaining how this is a demo explaining multiple content return types if there isn't one already and how it picks up on the tags

File: outputs/openapi/schema.go [hunk 8]
```diff
@@ -5,63 +5,80 @@ import (
...
+			}, true
 		}
+
+		return Schema{}, false
 	} else {
 		schema := mapAcceptedType(field.Type)
 		if field.Type == "slice" {
 			itemSchema := Schema{
-				Type: mapAcceptedType(field.SliceType).Type,
+				Type: mapAcceptedType(field.Type).Type,
```
Comment (Will-Mann-16) [inline outputs/openapi/schema.go:58]: Isn't field.type here slice? So shouldn't this not work

Comment (Will-Mann-16) [reply outputs/openapi/schema.go:58]: Unless it is array and then the items handle it lower down the line?

**Review by Will-Mann-16 — CHANGES_REQUESTED** [review]

File: examples/with-gorm/openapi.generated.yaml (hunk 6 above)
Comment (TommoLeedsy) [reply examples/with-gorm/openapi.generated.yaml]: This has now been updated, it was better but I have now added a fallback back if no binding tags are used.

File: examples/with-cli/openapi.generated.yaml (hunk 7 above)
Comment (TommoLeedsy) [reply examples/with-cli/openapi.generated.yaml]: Accidentally, it has now been put back.

File: outputs/openapi/schema.go (hunk 8 above)
Comment (TommoLeedsy) [reply outputs/openapi/schema.go]: I think I must have changed it during testing, I have reverted it and re run some of the examples and they don't change. I think the variable get overwritten anyway. We really need to improve our testing.

File: examples/with-multi-content-types/deletePost.go (hunk 1 above)
Comment (TommoLeedsy) [reply examples/with-multi-content-types/deletePost.go]: I have removed the surrounding curly braces on the cases which I think has improved the look of the switch?

File: examples/with-multi-content-types/main.go (hunk 2 above)
Comment (TommoLeedsy) [reply examples/with-multi-content-types/main.go:18]: I have now added that explanation to the readme

**Review by Will-Mann-16 — APPROVED** [review]

File: outputs/openapi/generate.go (hunk 5 above)
Comment (Will-Mann-16) [reply outputs/openapi/generate.go:161]: It's a nit pick, doesn't affect functionality at all

File: examples/with-multi-content-types/deletePost.go (hunk 1 above)
Comment (Will-Mann-16) [reply examples/with-multi-content-types/deletePost.go]: Cool, think it looks fine now.
